lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DoctrineCodingStandard\Sniffs\Classes;
+
+use DoctrineCodingStandard\Helpers\UseStatementHelper;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use const T_ABSTRACT;
+use const T_CLASS;
+use const T_FINAL;
+use function count;
+use function preg_grep;
+use function preg_match;
+
+final class ExceptionClassNamingSniff implements Sniff
+{
+    private const CODE_NOT_AN_EXCEPTION_CLASS = 'NotAnExceptionClass';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register() : array
+    {
+        return [T_CLASS];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr) : void
+    {
+        $isAbstract              = $phpcsFile->findFirstOnLine([T_ABSTRACT], $stackPtr) !== false;
+        $isFinal                 = $phpcsFile->findFirstOnLine([T_FINAL], $stackPtr) !== false;
+        $isExtendingException    = $this->isExtendingException($phpcsFile, $stackPtr);
+        $isImplementingException = $this->isImplementingException($phpcsFile, $stackPtr);
+        $hasExceptionName        = $this->hasExceptionSuffix((string) $phpcsFile->getDeclarationName($stackPtr));
+        $hasValidClassName       = ($isAbstract && $hasExceptionName) ||
+            (! $isAbstract && ! $hasExceptionName && $isFinal);
+        $isValidException        = $hasValidClassName && ($isExtendingException || $isImplementingException);
+        $isNoException           = ! $hasExceptionName && ! $isExtendingException && ! $isImplementingException;
+
+        if ($isValidException || $isNoException) {
+            return;
+        }
+
+        if (! $hasValidClassName) {
+            $phpcsFile->addError(
+                'Use "Exception" suffix for abstract exception classes and make non-abstract classes final',
+                $stackPtr,
+                self::CODE_NOT_AN_EXCEPTION_CLASS
+            );
+
+            return;
+        }
+
+        $phpcsFile->addError(
+            'Class is not a valid exception',
+            $stackPtr,
+            self::CODE_NOT_AN_EXCEPTION_CLASS
+        );
+    }
+
+    private function isExtendingException(File $phpcsFile, int $stackPtr) : bool
+    {
+        $extendsClass = $phpcsFile->findExtendedClassName($stackPtr);
+        if ($extendsClass === false) {
+            return false;
+        }
+
+        return $this->hasExceptionSuffix($extendsClass);
+    }
+
+    private function isImplementingException(File $phpcsFile, int $stackPtr) : bool
+    {
+        $implementedInterfaces = $phpcsFile->findImplementedInterfaceNames($stackPtr);
+        if ($implementedInterfaces === false) {
+            return false;
+        }
+
+        $isImplementingExceptions = count(preg_grep('/Exception$/', $implementedInterfaces)) > 0;
+
```

Comment:
You should early-exit here if `$isImplementingExceptions` is true.

Reply:
This is the sniff for classes. Implementing an Exception interface is not enough for a class, because it needs to extend another Exception class too. I don't know about a requirement that an Exception class has to implement an interface.

Reply:
>  Implementing an Exception interface is not enough for a class

That's true, but this sniff is about naming only, it should not report inheritance issues IMO.
Not extending some Exception will cause PHP fatal error, but that is out of scope for this sniff.

Reply:
I even start to see no point here of checking for an exception interface when a class, regular or abstract, has to extend an exception. This is a relict from before making exception classes `final`. I think the complexity of checks for exception interfaces can be removed here. How do you think about that?